File: mobile/src/core/care/ecoCoach.ts
import type { VehicleSnapshot } from "../obd/vehicleSnapshot";
import type { CareCue, CareContext } from "./careTypes";
import type { CareFeature } from "./careFeature";

class EcoCoach implements CareFeature {
  id = "ecoCoach";

  liveScore = 100;

  private speedHistory: Array<{ t: number; v: number }> = [];
  private accelStreakStart: number | undefined;
  private highRevCruiseStart: number | undefined;
  private lastThrottleOpenAt: number | undefined;
  private overrunStart: number | undefined;
  private idleSince: number | undefined;
  private highSpeedSince: number | undefined;
  private stopCountWindow: number[] = [];
  private wasStopped = false;
  private scoreWindow: Array<{ t: number; delta: number }> = [];

  isEnabled(settings: { careEcoCoach: boolean }): boolean {
    return settings.careEcoCoach;
  }

  evaluate(snapshot: VehicleSnapshot, context: CareContext): CareCue[] {
    if (context.isColdPhase) {
      context.liveEcoScore = this.liveScore;
      return [];
    }

    const now = context.now;
    const speed = snapshot.speedKmh ?? 0;
    const rpm = snapshot.rpm ?? 0;
    const throttle = snapshot.throttlePct ?? 0;
    const fuel = snapshot.engineFuelRateLh ?? 0;

    this.speedHistory.push({ t: now, v: speed });
    this.speedHistory = this.speedHistory.filter((s) => now - s.t <= 60_000);

    let accel = 0;
    if (this.speedHistory.length >= 2) {
      const a = this.speedHistory[this.speedHistory.length - 2]!;
      const b = this.speedHistory[this.speedHistory.length - 1]!;
      const dt = (b.t - a.t) / 1000;
      if (dt > 0) accel = (b.v - a.v) / 3.6 / dt;
    }

    const stopped = speed < 2;
    if (stopped && !this.wasStopped) this.stopCountWindow.push(now);
    this.wasStopped = stopped;
    this.stopCountWindow = this.stopCountWindow.filter((t) => now - t <= 300_000);
    const avgSpeed = this.speedHistory.length ? this.speedHistory.reduce((a, s) => a + s.v, 0) / this.speedHistory.length : speed;
    const isCityTraffic = this.stopCountWindow.length >= 8 && avgSpeed < 15;
    context.isCityTraffic = isCityTraffic;

    const cues: CareCue[] = [];
    const warn = (cue: CareCue) => {
      cues.push(cue);
      this.scoreWindow.push({ t: now, delta: -8 });
    };

    if (accel > 2.2) {
      if (this.accelStreakStart == null) this.accelStreakStart = now;
      if (now - this.accelStreakStart >= 1500) {
        warn({ id: "coach.easeOff", text: "Ease off the accelerator — smoother inputs save fuel.", severity: "coach" });
        this.accelStreakStart = now + 5000;
      }
    } else {
      this.accelStreakStart = undefined;
    }

    if (!isCityTraffic) {
      const vals = this.speedHistory.map((s) => s.v);
      const speedStable = this.speedHistory.length >= 4 && Math.max(...vals) - Math.min(...vals) < 3;
      if (rpm > 2800 && speedStable && speed > 20) {
        if (this.highRevCruiseStart == null) this.highRevCruiseStart = now;
        if (now - this.highRevCruiseStart >= 8000) {
          warn({ id: "coach.shiftUp", text: "Shift up — you're cruising at higher RPM than needed.", severity: "coach" });
          this.highRevCruiseStart = now + 20_000;
        }
      } else {
        this.highRevCruiseStart = undefined;
      }
    }

    if (throttle > 5) this.lastThrottleOpenAt = now;
    if (accel < -2.8 && this.lastThrottleOpenAt != null && now - this.lastThrottleOpenAt < 5000) {
      warn({ id: "coach.liftEarlier", text: "Try lifting off the throttle a bit earlier before braking.", severity: "coach" });
      this.lastThrottleOpenAt = undefined;
    }

    if (speed > 25 && throttle > 5 && accel < -0.5 && fuel > 0.5) {
      warn({ id: "coach.coast", text: "Coast instead of holding throttle while slowing down.", severity: "coach" });
    }

    if (speed > 25 && throttle < 2 && fuel < 0.1) {
      if (this.overrunStart == null) this.overrunStart = now;
      if (now - this.overrunStart >= 4000) {
        cues.push({ id: "coach.nice", text: "Nice coasting — that saves fuel.", severity: "celebration" });
        this.scoreWindow.push({ t: now, delta: 6 });
        this.overrunStart = now + 30_000;
      }
    } else {
      this.overrunStart = undefined;
    }

    if (!isCityTraffic && speed > 70) {
      if (speedOscillations(this.speedHistory, 8) >= 4) {
        warn({ id: "coach.steady", text: "Try to hold a steadier speed on the highway.", severity: "coach" });
      }
    }

    if (speed < 2 && rpm > 300) {
      if (this.idleSince == null) this.idleSince = now;
      if (now - this.idleSince > 180_000) {
        warn({ id: "coach.longIdle", text: "Consider turning off the engine during long idles.", severity: "coach" });
        this.idleSince = now + 300_000;
      }
    } else {
      this.idleSince = undefined;
    }

    if (!isCityTraffic && speed > 120) {
      if (this.highSpeedSince == null) this.highSpeedSince = now;
      if (now - this.highSpeedSince >= 20_000) {
        warn({ id: "coach.highSpeed", text: "High speed increases fuel consumption significantly.", severity: "coach" });
        this.highSpeedSince = now + 60_000;
      }
    } else {
      this.highSpeedSince = undefined;
    }

    this.updateLiveScore(now);
    context.liveEcoScore = this.liveScore;
    return cues;
  }

  private updateLiveScore(now: number) {
    this.scoreWindow = this.scoreWindow.filter((s) => now - s.t <= 30_000);
    const delta = this.scoreWindow.reduce((a, s) => a + s.delta, 0);
    this.liveScore = Math.min(100, Math.max(0, 100 + delta));
  }

  resetTrip(): void {
    this.speedHistory = [];
    this.liveScore = 100;
    this.scoreWindow = [];
  }
}

function speedOscillations(history: Array<{ t: number; v: number }>, amplitude: number): number {
  if (history.length < 6) return 0;
  let crossings = 0;
  const mean = history.reduce((a, s) => a + s.v, 0) / history.length;
  let above = history[0]!.v > mean;
  for (let i = 1; i < history.length; i++) {
    const v = history[i]!.v;
    const nowAbove = v > mean + amplitude / 2;
    const nowBelow = v < mean - amplitude / 2;
    if (above && nowBelow) {
      crossings++;
      above = false;
    }
    if (!above && nowAbove) {
      crossings++;
      above = true;
    }
  }
  return crossings;
}

export const ecoCoach = new EcoCoach();

File: QuickCar/Core/Care/EcoCoach.swift
import Foundation

@MainActor
final class EcoCoach: CareFeature {
    let id = "ecoCoach"
    let requiredPIDs: Set<UInt8> = [0x0D, 0x0C]
    let optionalPIDs: Set<UInt8> = [0x10, 0x5E, 0x11]

    private var speedHistory: [(Date, Double)] = []
    private var accelStreakStart: Date?
    private var highRevCruiseStart: Date?
    private var lastThrottleOpenAt: Date?
    private var coastCandidateStart: Date?
    private var overrunStart: Date?
    private var idleSince: Date?
    private var highSpeedSince: Date?
    private var stopCountWindow: [(Date, Bool)] = []
    private var wasStopped = false
    private var warningCount = 0
    private var scoreWindow: [(Date, Double)] = []

    private(set) var liveScore: Double = 100

    func isEnabled(settings: AppSettings) -> Bool { settings.careEcoCoach }

    func evaluate(snapshot: VehicleSnapshot, context: inout CareContext) -> [CareCue] {
        guard !context.isColdPhase else {
            context.liveEcoScore = liveScore
            return []
        }

        let now = context.now
        let speed = snapshot.speedKmh ?? 0
        let rpm = snapshot.rpm ?? 0
        let throttle = snapshot.throttlePct ?? 0
        let fuel = snapshot.engineFuelRateLh ?? 0

        speedHistory.append((now, speed))
        speedHistory = speedHistory.filter { now.timeIntervalSince($0.0) <= 60 }

        // Acceleration from speed delta
        var accel = 0.0
        if speedHistory.count >= 2 {
            let a = speedHistory[speedHistory.count - 2]
            let b = speedHistory[speedHistory.count - 1]
            let dt = b.0.timeIntervalSince(a.0)
            if dt > 0 {
                accel = ((b.1 - a.1) / 3.6) / dt
            }
        }

        let stopped = speed < 2
        if stopped, !wasStopped {
            stopCountWindow.append((now, true))
        }
        wasStopped = stopped
        stopCountWindow = stopCountWindow.filter { now.timeIntervalSince($0.0) <= 300 }
        let avgSpeed5m: Double = {
            let w = speedHistory // reuse 60s; approximate with recent
            let longer = speedHistory
            return longer.isEmpty ? speed : longer.map(\.1).reduce(0, +) / Double(longer.count)
        }()
        // City traffic: use stop count + low speed heuristic over rolling buffer
        context.isCityTraffic = stopCountWindow.count >= 8 && avgSpeed5m < 15

        var cues: [CareCue] = []
        func warn(_ cue: CareCue) {
            warningCount += 1
            cues.append(cue)
            scoreWindow.append((now, -8))
        }

        // Hard accel
        if accel > 2.2 {
            if accelStreakStart == nil { accelStreakStart = now }
            if let s = accelStreakStart, now.timeIntervalSince(s) >= 1.5 {
                warn(CareCue(id: "coach.easeOff", text: LocalizedCare.string("coach.easeOff"), severity: .coach, localizationKey: "coach.easeOff"))
                accelStreakStart = now.addingTimeInterval(5)
            }
        } else {
            accelStreakStart = nil
        }

        if !context.isCityTraffic {
            let speedStable = speedHistory.count >= 4
                && (speedHistory.map(\.1).max()! - speedHistory.map(\.1).min()!) < 3
            if rpm > 2800, speedStable, speed > 20 {
                if highRevCruiseStart == nil { highRevCruiseStart = now }
                if let s = highRevCruiseStart, now.timeIntervalSince(s) >= 8 {
                    warn(CareCue(id: "coach.shiftUp", text: LocalizedCare.string("coach.shiftUp"), severity: .coach, localizationKey: "coach.shiftUp"))
                    highRevCruiseStart = now.addingTimeInterval(20)
                }
            } else {
                highRevCruiseStart = nil
            }
        }

        if throttle > 5 { lastThrottleOpenAt = now }
        if accel < -2.8 {
            if let open = lastThrottleOpenAt, now.timeIntervalSince(open) < 5 {
                warn(CareCue(id: "coach.liftEarlier", text: LocalizedCare.string("coach.liftEarlier"), severity: .coach, localizationKey: "coach.liftEarlier"))
                lastThrottleOpenAt = nil
            }
        }

        if speed > 25, throttle < 2, fuel < 0.15, accel < 0 {
            if coastCandidateStart == nil { coastCandidateStart = now }
            // Missed coast if throttle was open while decelerating with fuel
        } else if speed > 25, throttle > 5, accel < -0.5, fuel > 0.5 {
            warn(CareCue(id: "coach.coast", text: LocalizedCare.string("coach.coast"), severity: .coach, localizationKey: "coach.coast"))
        } else {
            coastCandidateStart = nil
        }

        // Overrun reward
        if speed > 25, throttle < 2, fuel < 0.1 {
            if overrunStart == nil { overrunStart = now }
            if let s = overrunStart, now.timeIntervalSince(s) >= 4 {
                cues.append(CareCue(id: "coach.nice", text: LocalizedCare.string("coach.nice"), severity: .celebration, localizationKey: "coach.nice"))
                scoreWindow.append((now, 6))
                overrunStart = now.addingTimeInterval(30)
            }
        } else {
            overrunStart = nil
        }

        if !context.isCityTraffic, speed > 70 {
            let osc = Self.speedOscillations(speedHistory, amplitude: 8)
            if osc >= 4 {
                warn(CareCue(id: "coach.steady", text: LocalizedCare.string("coach.steadySpeed"), severity: .coach, localizationKey: "coach.steadySpeed"))
            }
        }

        if speed < 2, rpm > 300 {
            if idleSince == nil { idleSince = now }
            if let s = idleSince, now.timeIntervalSince(s) > 180 {
                warn(CareCue(id: "coach.longIdle", text: LocalizedCare.string("coach.longIdle"), severity: .coach, localizationKey: "coach.longIdle"))
                idleSince = now.addingTimeInterval(300)
            }
        } else {
            idleSince = nil
        }

        if !context.isCityTraffic {
            if speed > 120 {
                if highSpeedSince == nil { highSpeedSince = now }
                if let s = highSpeedSince, now.timeIntervalSince(s) >= 20 {
                    warn(CareCue(id: "coach.highSpeed", text: LocalizedCare.string("coach.highSpeed"), severity: .coach, localizationKey: "coach.highSpeed"))
                    highSpeedSince = now.addingTimeInterval(60)
                }
            } else {
                highSpeedSince = nil
            }
        }

        updateLiveScore(now: now)
        context.liveEcoScore = liveScore
        return cues
    }

    private func updateLiveScore(now: Date) {
        scoreWindow = scoreWindow.filter { now.timeIntervalSince($0.0) <= 30 }
        let delta = scoreWindow.map(\.1).reduce(0, +)
        liveScore = min(100, max(0, 100 + delta))
    }

    static func speedOscillations(_ history: [(Date, Double)], amplitude: Double) -> Int {
        guard history.count >= 6 else { return 0 }
        var crossings = 0
        let mean = history.map(\.1).reduce(0, +) / Double(history.count)
        var above = history[0].1 > mean
        for s in history.dropFirst() {
            let nowAbove = s.1 > mean + amplitude / 2
            let nowBelow = s.1 < mean - amplitude / 2
            if above, nowBelow { crossings += 1; above = false }
            if !above, nowAbove { crossings += 1; above = true }
        }
        return crossings
    }

    func resetTrip() {
        speedHistory.removeAll()
        warningCount = 0
        liveScore = 100
        scoreWindow.removeAll()
    }
}
